### ISSUE TITLE
Add implicit mount deps for units with ConditionPathIsDirectory= and similar conditions

### ISSUE THREAD

**Issue by kakra** [systemd/systemd#3727]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> 229

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> Gentoo
### In case of bug report: Expected behaviour you didn't see

> ConditionPathIsDirectory=/mnt/automounted-device/some-directory should trigger automount and detect if the directory exists
### In case of bug report: Unexpected behaviour you saw

> ConditionPathIsDirectory didn't trigger the automount, it simply said in the log that the condition wasn't met. As soon as I cd'ed into the directory with bash and retried, the condition was met upon starting the service.
### In case of bug report: Steps to reproduce the problem

> Setup an automount point, setup a service with ConditionPathIsDirectory pointing to a child directory of the automounted device, start the service:

```
$ systemctl cat mnt-private-usb\\x2dbackup.{mount,automount} usb-backup.{timer,service}
# /run/systemd/generator/mnt-private-usb\x2dbackup.mount
# Automatically generated by systemd-fstab-generator

[Unit]
SourcePath=/etc/fstab
Documentation=man:fstab(5) man:systemd-fstab-generator(8)

[Mount]
What=/dev/disk/by-label/usb-backup
Where=/mnt/private/usb-backup
Type=ext4
Options=noauto,noatime,x-systemd.automount,x-systemd.idle-timeout=300

# /run/systemd/generator/mnt-private-usb\x2dbackup.automount
# Automatically generated by systemd-fstab-generator

[Unit]
SourcePath=/etc/fstab
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
Before=local-fs.target
[Automount]
Where=/mnt/private/usb-backup
TimeoutIdleSec=5min

# /etc/systemd/system/usb-backup.timer
[Unit]
Description=Daily USB Backup Timer

[Timer]
OnCalendar=03:00
WakeSystem=true

[Install]
WantedBy=timers.target

# /etc/systemd/system/usb-backup.service
[Unit]
Description=Daily USB Backup Service
ConditionPathIsDirectory=/mnt/private/usb-backup/jupiter.sol.local.borg

[Service]
Type=oneshot
WorkingDirectory=/mnt/private/usb-backup
ExecStart=/usr/bin/borg create --stats --exclude-caches --compression lz4 /mnt/private/usb-backup/jupiter.sol.local.borg::'system@{now:%%Y%%m%%d-%%H%%M}' /mnt/btrfs-pool
IOSchedulingClass=idle
IOSchedulingPriority=7
CPUSchedulingPolicy=batch
Nice=3
PrivateTmp=true
ProtectHome=read-only
ProtectSystem=full
ReadWriteDirectories=/root/.cache/borg
```

**Comment by evverx** [comment]
Indeed, `systemd` is the automount daemon. So, our `stat` doesn't trigger automount and we get

```
stat("/some/path/automount-point/some-dir", 0x7ffcfc9f4ea0) = -1 ENOENT (No such file or directory)
```

**Comment by poettering** [comment]
I figure we could add implicit ordering for all paths specified in conditions, so that starting a service with a condition on a path will always implicitly pull in the mounts necessary for the ultimate condition check first.

I mean, we already add similar implicit deps for WorkingDirectory= and RootDirectory=, so this would be a pretty natural extension.

**Comment by evverx** [comment]
But what about mountpoint expiry? Is this scenario possible:
- we succefully check a mountpoint
- mountpoint expires
- our `stat` fails

?

**Comment by kakra** [comment]
@poettering From my example you can see that I already set WorkingDirectory... But it also didn't trigger the automount (I think it instead froze and printed a strange message to dmesg). But that may well be another bug I didn't try to reproduce yet. Of course, it probably doesn't work that way anyways (I guess the condition comes before cwd).
